inspirehep/alembic/0bc0a6ee1bc0_create_index_on_referenced_records.py

```diff
@@ -0,0 +1,56 @@
+#
+# This file is part of Invenio.
+# Copyright (C) 2016-2018 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""create index on referenced_records"""
+from __future__ import absolute_import, division, print_function
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = '0bc0a6ee1bc0'
+down_revision = '2f5368ff6d20'
+branch_labels = ()
+depends_on = '07fb52561c5c'
+
+
+def upgrade():
+    """Upgrade database."""
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.execute('''
+        create or replace function referenced_records(json jsonb) returns text[] as  $$
+        select
+            array_agg(
+              split_part(
+                reference_arr->'record'->>'$ref',E'literature/',2
+              )
+            )  as result
+         from
+            jsonb_array_elements(json->'references') as reference_arr; $$
+        language sql IMMUTABLE
+    ''')
+    op.execute('''
+        create index ix_records_metadata_json_referenced_records
+            on records_metadata
+            using gin(referenced_records(json))
+    ''')
+
+    op.execute('''
+        create or replace function count_citations(ids text) returns bigint as $$
+        select count(id) as citations_count from records_metadata
+            where referenced_records(json) @> ARRAY[ids]; $$
```

Comment:
this is incorrect, as it's not [`immutable`](https://www.postgresql.org/docs/9.4/static/xfunc-volatility.html). Besides, I am not sure hardcoding this in the DB is the best way to go, as we will need to remove some specific records from the counts in the future, based on metadata of the citing/cited record. If you do this in Python, you can also drop the string filtering in the DB and have the `referenced_records` return the full URI.